hoomd/data/local_access.py

```diff
@@ -0,0 +1,453 @@
+from abc import ABC, abstractmethod
+from hoomd import Box
+from hoomd.array import HOOMDArray
+from hoomd import _hoomd
+
+
+class _LocalAccess(ABC):
+    @property
+    @abstractmethod
+    def _fields(self):
+        pass
+
+    _global_fields = {'rtag': 'getRTags'}
+
+    @property
+    @abstractmethod
+    def _array_cls(self):
+        pass
+
+    def __init__(self):
+        object.__setattr__(self, '_entered', False)
+        object.__setattr__(self, '_accessed_fields', dict())
+
+    def __getattr__(self, attr):
+        if attr in self._accessed_fields:
+            return self._accessed_fields[attr]
+        elif attr in self._global_fields:
+            buff = getattr(self._cpp_obj, self._global_fields[attr])()
+        elif attr.startswith("ghost_"):
+            if attr[6:] in self._fields:
+                buff = getattr(self._cpp_obj, self._fields[attr[6:]])(
+                    True, False)
+            else:
+                raise AttributeError(
+                    "{} object has no attribute {}".format(type(self), attr))
+        elif attr.endswith("_with_ghosts"):
+            if attr[:-12] in self._fields:
+                buff = getattr(self._cpp_obj, self._fields[attr[:-12]])(
+                    False, True)
+            else:
+                raise AttributeError(
+                    "{} object has no attribute {}".format(type(self), attr))
+        elif attr in self._fields:
+            buff = getattr(self._cpp_obj, self._fields[attr])(False, False)
+        else:
+            raise AttributeError(
+                "{} object has no attribute {}".format(type(self), attr))
+        arr = self._array_cls(buff, lambda: self._entered)
+        self._accessed_fields[attr] = arr
+        return arr
+
+    def __setattr__(self, attr, value):
+        try:
+            arr = getattr(self, attr)
+        except AttributeError:
+            raise AttributeError(
+                "Cannot set attribute {}, does not exist.".format(attr))
+        if arr.read_only == True:
+            raise RuntimeError(
+                "Attribute {} is not settable.".format(attr))
+        arr[:] = value
```

Comment:
That is not how classes behave by default no. I was following how the `Snapshot` object currently behaves (although the same argument can be made there though the attributes aren't directly settable). Also, since `dir` or `__dict__` won't show the available options, I wanted to help users not accidentally type the wrong thing. Also, while it is not generally Pythonic, `numpy.ndarray` and all standard Python classes do not let users set arbitrary attributes.

I don't feel strongly either way, I admit it is difficult to remember to always use `object.__setattr__` for setting things even internally.

Reply:
That's fair, if you want to do that then I think a more Pythonic solution might be to use `__slots__` and then accept my change. The call to `super().__setattr__` should then handle preventing assignment as needed.

Reply:
I like that solution. However, I don't want to just get my way, necessarily, as a user would you want to be able to set attributes arbitrarily rather than have stronger guarantees that you are assigning what you think you are?

Reply:
That's a tricky question in this specific case. IMO the standard way for handling this kind of thing (for example in fresnel, OVITO, matplotlib, etc) is to require users to handle the `[:]` syntax themselves, i.e. users should do `snap.particles.position[:] = ...`. I think trying to get Python to behave like a compiled, less dynamic language is generally inadvisable because it requires jumping through too many hoops, leads to code that is harder to read and maintain, and doesn't really follow the Zen of Python.

Having said that, in this case you're not using a numpy array so much as you're directly trying to mirror numpy array syntax, so I can see a good argument for doing it. Regarding Python built-in classes I think the reasoning has more to do with the default behavior of C extension classes, but I can see wanting to match numpy very closely here. So I think it's fine to do this using the slots solution, which at least helps avoid a large amount of internal developer error.